src/dune_engine/sandbox.ml

```diff
@@ -166,7 +214,47 @@ let rename_dir_recursively ~loc ~src_dir ~dst_dir =
   in
   loop ~src_dir ~dst_dir |> Path.Build.Set.of_list
 
+let apply_changes_to_source_tree t ~old_snapshot =
+  let new_snapshot = snapshot t in
+  (* Same as promotion: make the file writable when copying to the source
+     tree. *)
+  let in_source_tree p =
+    Path.extract_build_context_dir_maybe_sandboxed p
+    |> Option.value_exn |> snd |> Path.source
+  in
+  let copy_file p =
+    let in_source_tree = in_source_tree p in
+    Path.unlink_no_err in_source_tree;
+    Option.iter (Path.parent in_source_tree) ~f:Path.mkdir_p;
+    Io.copy_file ~src:p ~dst:in_source_tree ()
+  in
+  let delete_file p =
+    let in_source_tree = in_source_tree p in
+    Path.unlink_no_err in_source_tree
+  in
+  let (_ : _ Path.Map.t) =
+    Path.Map.merge old_snapshot new_snapshot ~f:(fun p before after ->
+        match (before, after) with
+        | None, None -> assert false
+        | None, Some _ ->
+          copy_file p;
+          None
```

Comment:
This return of `None` looks a bit weird but perhaps there is no `merge`-like function that allows you to avoid creating the resulting map. If we are forced to use `None` everywhere, perhaps, it's worth asserting that the resulting map is empty.

Reply:
I added `Map.iter2`

Reply:
Nice, thanks!